docs/specs/alcproxies.md

```diff
@@ -0,0 +1,279 @@
+* Build proxies for each non-simple type that can communicate across the ALC boundaries
+* Make everything a shared type (This is not recommended)
+* Encode any passed objects and decode them on the other side, creating an effective "copy" of the type in the form `ALC.Foo`
+
+The first two have their own issues, as we would either need to figure out how to build those proxies on the fly from the IL-Generated `ProxyObject`/`ServerObject`, or lose out on the benefit of versioning compatibility. The third option is what `ServerObject` and `ClientObject` intend on setting a framework for; allowing for a (possibly interchangeable) connection framework to be used to send messages to and from the ALCs (or processes).
+
+There are a few ways we could implement the client/server encoding and messaging. The few that I've been looking at are:
+* Add a second `DispatchProxy` that can call across the barrier and keep our abstraction layer.
+* Using `BinaryFormatter` to serialize data across the boundary
+* Similar serialization with `DataContractSerializer`
+* Using gRPC to send objects across
+
+#### Second `DispatchProxy`
+Making the `ClientObject` a `DispatchProxy` object would work for making calls over ALCs, though not for out-of-process calls. It's not the perfect solution, as it would require two `Reflection.Invoke` calls (one from client to server, and one from server to target), but it would allow us to enforce our isolation policy from the custom `ServerObject` being able to mess with our return calls to make sure we were enforcing our constraints (see "ALC Unloading" below).
+
+#### Binary Serialization
+Using `BinaryFormatter`, types and their information can be serialized and sent across the ALC barrier, and deserialized using a class for decoding, which can then call the specific `TargetObject` methods.
+
+While Binary Serialization in concept is a good idea, there are too many issues that make this an ineffective solution. `BinaryFormatter` is a depricated API, many types aren't serializable, limiting the use of types we could proxy. 
+
+#### XML/JSON Serialization using `DataContractSerializer`
+Instead of moving a specific type through a `BinaryFormatter`, using `DataContractSerializer` would allow us to move any public or private pieces of an object, and then recreate it on the other side.
+
+Benefits of `DataContractSerializer`:
+* It may be easier compared to `BinaryFormatter` to cast this effectively across the type-isolation barrier.
+* Adding attributes is similar to `BinaryFormatter`, so any type should be able to make it across the barrier.
+
+Concerns with `DataContractSerializer`:
+* Many similar concerns to `BinaryFormatter`, such as potential security issues and the manual use of [DataContractAttribute] and [DataMemberAttribute] needing to be used. There may also be limits to what types can be serialized using `DataContractSerializer`.
```

Comment:
I put some of this below but want to make sure it's contextually in the right place.  
1. There aren't any potential security issues with DataContractSerializer. The issue with BinaryFormatter is that it allows the serialized payload to dictate what types are instantiated. This is NOT the case with DataContractSerializer.  
2. DataContractSerializer does NOT require usage of attributes to be able to serialize. The default is to serialize all public members. If you apply the DataContractAttribute, it will only serialize members which have the DataMemberAttribute applied. But if you want everything serialized, then you don't need any attributes at all. It also supports the SerializableAttribute, ISerializable interface and I think IXmlSerializable interface (but that might be limited to XmlSerializer, I'd have to check) to control serialization.